File: go/internal/cmd/commit.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	commitPending bool
	commitMsg     string
	commitAuto    bool
)

var commitCmd = &cobra.Command{
	Use:   "commit [ids...]",
	Short: "Commit thread changes",
	Long: `Commit specific threads or all pending thread changes.

Use --pending to commit all modified threads at once.`,
	RunE: runCommit,
}

func init() {
	commitCmd.Flags().BoolVar(&commitPending, "pending", false, "Commit all modified threads")
	commitCmd.Flags().StringVarP(&commitMsg, "m", "m", "", "Commit message")
	commitCmd.Flags().BoolVar(&commitAuto, "auto", false, "Auto-accept generated message")
}

func runCommit(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	var files []string

	if commitPending {
		// Collect all thread files with uncommitted changes
		threads, err := workspace.FindAllThreads(ws)
		if err != nil {
			return err
		}

		for _, t := range threads {
			relPath, _ := filepath.Rel(ws, t)
			if git.HasChanges(ws, relPath) {
				files = append(files, t)
			}
		}
	} else {
		// Resolve provided IDs to files
		if len(args) == 0 {
			return fmt.Errorf("provide thread IDs or use --pending")
		}

		for _, id := range args {
			file, err := workspace.FindByRef(ws, id)
			if err != nil {
				return err
			}
			relPath, _ := filepath.Rel(ws, file)
			if !git.HasChanges(ws, relPath) {
				fmt.Printf("No changes in thread: %s\n", id)
				continue
			}
			files = append(files, file)
		}
	}

	if len(files) == 0 {
		fmt.Println("No threads to commit.")
		return nil
	}

	// Generate commit message if not provided
	msg := commitMsg
	if msg == "" {
		msg = git.GenerateCommitMessage(ws, files)
		fmt.Printf("Generated message: %s\n", msg)

		if !commitAuto && isTerminal() {
			reader := bufio.NewReader(os.Stdin)
			fmt.Print("Proceed? [Y/n] ")
			response, _ := reader.ReadString('\n')
			response = strings.TrimSpace(strings.ToLower(response))
			if response == "n" || response == "no" {
				fmt.Println("Aborted.")
				return nil
			}
		}
	}

	// Stage and commit
	var relPaths []string
	for _, f := range files {
		relPath, _ := filepath.Rel(ws, f)
		relPaths = append(relPaths, relPath)
	}

	if err := git.Commit(ws, relPaths, msg); err != nil {
		return err
	}

	if err := git.Push(ws); err != nil {
		fmt.Printf("WARNING: git push failed (commit succeeded): %v\n", err)
	}

	fmt.Printf("Committed %d thread(s).\n", len(files))
	return nil
}

func isTerminal() bool {
	stat, _ := os.Stdin.Stat()
	return (stat.Mode() & os.ModeCharDevice) != 0
}

File: rust/src/cmd/commit.rs
use std::io::{self, BufRead, Write};
use std::path::Path;

use clap::Args;

use crate::git;
use crate::workspace;

#[derive(Args)]
pub struct CommitArgs {
    /// Thread IDs to commit
    ids: Vec<String>,

    /// Commit all modified threads
    #[arg(long)]
    pending: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,

    /// Auto-accept generated message
    #[arg(long)]
    auto: bool,
}

pub fn run(args: CommitArgs, ws: &Path) -> Result<(), String> {
    let mut files = Vec::new();

    if args.pending {
        // Collect all thread files with uncommitted changes
        let threads = workspace::find_all_threads(ws)?;

        for t in threads {
            let rel_path = t
                .strip_prefix(ws)
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_else(|_| t.to_string_lossy().to_string());

            if git::has_changes(ws, &rel_path) {
                files.push(t.to_string_lossy().to_string());
            }
        }
    } else {
        // Resolve provided IDs to files
        if args.ids.is_empty() {
            return Err("provide thread IDs or use --pending".to_string());
        }

        for id in &args.ids {
            let file = workspace::find_by_ref(ws, id)?;
            let rel_path = file
                .strip_prefix(ws)
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_else(|_| file.to_string_lossy().to_string());

            if !git::has_changes(ws, &rel_path) {
                println!("No changes in thread: {}", id);
                continue;
            }
            files.push(file.to_string_lossy().to_string());
        }
    }

    if files.is_empty() {
        println!("No threads to commit.");
        return Ok(());
    }

    // Generate commit message if not provided
    let msg = if let Some(m) = args.m {
        m
    } else {
        let generated = git::generate_commit_message(ws, &files);
        println!("Generated message: {}", generated);

        if !args.auto && is_terminal() {
            print!("Proceed? [Y/n] ");
            io::stdout().flush().ok();

            let mut response = String::new();
            io::stdin().lock().read_line(&mut response).ok();
            let response = response.trim().to_lowercase();

            if response == "n" || response == "no" {
                println!("Aborted.");
                return Ok(());
            }
        }

        generated
    };

    // Stage and commit
    let rel_paths: Vec<String> = files
        .iter()
        .map(|f| {
            Path::new(f)
                .strip_prefix(ws)
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_else(|_| f.clone())
        })
        .collect();

    git::commit(ws, &rel_paths, &msg)?;

    if let Err(e) = git::push(ws) {
        eprintln!("WARNING: git push failed (commit succeeded): {}", e);
    }

    println!("Committed {} thread(s).", files.len());
    Ok(())
}

fn is_terminal() -> bool {
    use std::io::IsTerminal;
    io::stdin().is_terminal()
}
